operators/pkg/controller/elasticsearch/driver/upgrade.go

```diff
@@ -0,0 +1,305 @@
+	if err != nil {
+		return false, err
+	}
+	if !green {
+		log.Info("Skipping node rolling upgrade since cluster is not green", "namespace", es.Namespace, "name", es.Name)
+		return false, nil
+	}
+	return true, nil
+}
+
+// podUpgradeDone inspects the given pod and returns true if it was successfully upgraded.
+func podUpgradeDone(c k8s.Client, esState ESState, podRef types.NamespacedName, expectedRevision string) (bool, error) {
+	if expectedRevision == "" {
+		// no upgrade scheduled for the sset
+		return false, nil
+	}
+	// retrieve pod to inspect its revision label
+	var pod corev1.Pod
+	err := c.Get(podRef, &pod)
+	if err != nil && !errors.IsNotFound(err) {
+		return false, err
+	}
+	if errors.IsNotFound(err) || !pod.DeletionTimestamp.IsZero() {
+		// pod is terminating
+		return false, nil
+	}
+	if sset.PodRevision(pod) != expectedRevision {
+		// pod revision does not match the sset upgrade revision
+		return false, nil
+	}
+	// is the pod ready?
+	if !k8s.IsPodReady(pod) {
+		return false, nil
+	}
+	// has the node joined the cluster yet?
+	inCluster, err := esState.NodesInCluster([]string{podRef.Name})
+	if err != nil {
+		return false, err
+	}
+	if !inCluster {
+		log.V(1).Info("Node has not joined the cluster yet", "namespace", podRef.Namespace, "name", podRef.Name)
+		return false, err
+	}
+	return true, nil
+}
+
+func disableShardsAllocation(esClient esclient.Client) error {
+	ctx, cancel := context.WithTimeout(context.Background(), esclient.DefaultReqTimeout)
+	defer cancel()
+	return esClient.DisableReplicaShardsAllocation(ctx)
+}
+
+func doSyncFlush(esClient esclient.Client) error {
+	ctx, cancel := context.WithTimeout(context.Background(), esclient.DefaultReqTimeout)
+	defer cancel()
+	return esClient.SyncedFlush(ctx)
+}
+
+func (d *defaultDriver) MaybeEnableShardsAllocation(
+	es v1alpha1.Elasticsearch,
+	esClient esclient.Client,
+	esState ESState,
+	statefulSets sset.StatefulSetList,
+) *reconciler.Results {
+	results := &reconciler.Results{}
+	// Since we rely on sset rollingUpdate.Partition, requeue in case our cache hasn't seen a sset update yet.
+	// Otherwise we could re-enable shards allocation while a pod was just scheduled for termination,
+	// with the partition in the sset cache being outdated.
+	if !d.Expectations.GenerationExpected(statefulSets.ObjectMetas()...) {
+		return results.WithResult(defaultRequeue)
+	}
+
+	alreadyEnabled, err := esState.ShardAllocationsEnabled()
+	if err != nil {
+		return results.WithError(err)
+	}
+	if alreadyEnabled {
+		return results
+	}
+
+	// Make sure all pods scheduled for upgrade have been upgraded.
+	scheduledUpgradesDone, err := sset.ScheduledUpgradesDone(d.Client, statefulSets)
+	if err != nil {
+		return results.WithError(err)
+	}
+	if !scheduledUpgradesDone {
+		log.V(1).Info(
+			"Rolling upgrade not over yet, some pods don't have the updated revision, keeping shard allocations disabled",
+			"namespace", es.Namespace,
+			"name", es.Name,
```

Comment:
In #1168, we chose to be explicit with the owner of the name and log `"es_name", es.Name`.

Reply:
Yes! I'll merge that one and fix it in #1262.